components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/attribute/handler/MagicLinkAuthAttributeHandler.java

```diff
@@ -0,0 +1,114 @@
+import org.wso2.carbon.identity.auth.attribute.handler.model.AuthAttributeHolder;
+import org.wso2.carbon.identity.auth.attribute.handler.model.AuthAttributeType;
+import org.wso2.carbon.identity.auth.attribute.handler.model.ValidationFailureReason;
+import org.wso2.carbon.identity.auth.attribute.handler.model.ValidationResult;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static org.wso2.carbon.identity.auth.attribute.handler.AuthAttributeHandlerConstants.ErrorMessages.ERROR_CODE_ATTRIBUTE_NOT_FOUND;
+import static org.wso2.carbon.identity.auth.attribute.handler.AuthAttributeHandlerConstants.ErrorMessages.ERROR_CODE_ATTRIBUTE_VALUE_EMPTY;
+
+/**
+ * Auth attribute handler implementation for the MagicLink authenticator.
+ */
+public class MagicLinkAuthAttributeHandler implements AuthAttributeHandler {
+
+    private static final String HANDLER_NAME = "MagicLinkAuthAttributeHandler";
+    private static final String ATTRIBUTE_EMAIL = "http://wso2.org/claims/emailaddress";
+
+    @Override
+    public String getName() {
+
+        return HANDLER_NAME;
+    }
+
+    @Override
+    public AuthAttributeHandlerBindingType getBindingType() {
+
+        return AuthAttributeHandlerBindingType.AUTHENTICATOR;
+    }
+
+    @Override
+    public String getBoundIdentifier() {
+
+        return MagicLinkAuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+
+    @Override
+    public AuthAttributeHolder getAuthAttributeData() {
+
+        List<AuthAttribute> authAttributes = new ArrayList<>();
+        authAttributes.add(buildAuthAttribute(MagicLinkAuthenticatorConstants.USER_NAME, false));
+        authAttributes.add(buildAuthAttribute(ATTRIBUTE_EMAIL, true));
+
+        return new AuthAttributeHolder(
+                getName(),
+                getBindingType(),
+                getBoundIdentifier(),
+                authAttributes
+        );
+    }
+
+    @Override
+    public ValidationResult validateAttributes(Map<String, String> attributeMap) {
+
+        ValidationResult validationResult = new ValidationResult(true);
+
+        validateAttributeExistence(MagicLinkAuthenticatorConstants.USER_NAME, attributeMap, validationResult);
+        validateAttributeExistence(ATTRIBUTE_EMAIL, attributeMap, validationResult);
+
+        return validationResult;
+    }
+
+    private AuthAttribute buildAuthAttribute(String name, boolean isClaim) {
+
+        return new AuthAttribute(name, isClaim, false, AuthAttributeType.STRING);
+    }
+
+    private void validateAttributeExistence(String attribute, Map<String, String> attributeMap,
+                                            ValidationResult validationResult) {
+
+        if (attributeMap == null || attributeMap.isEmpty() || !attributeMap.containsKey(attribute)) {
+            validationResult.setValid(false);
+            addFailureReason(validationResult, attribute, ERROR_CODE_ATTRIBUTE_NOT_FOUND);
```

Comment:
If the attribute map is null, it will be added as failure reason against each attribute right?

Reply:
Yes. This is because the failure reason is kept against each attribute